Bottom: db6f379646ca0fae747ef5cd04b50c3952350431
Top:    03db0fc9f2cb988f6b56ccc07b7f9522e9f73c2f
Author: Sean Hefty <sean.hefty@intel.com>
Date:   2012-07-18 09:01:43 -0700

debug fork


---

diff --git a/src/preload.c b/src/preload.c
index 7be40f6..c83da39 100644
--- a/src/preload.c
+++ b/src/preload.c
@@ -48,6 +48,7 @@
 #include <netinet/tcp.h>
 #include <unistd.h>
 #include <semaphore.h>
+#include <stdio.h>
 
 #include <rdma/rdma_cma.h>
 #include <rdma/rdma_verbs.h>
@@ -466,18 +467,22 @@ static int connect_fork(int socket, const struct sockaddr *addr, socklen_t addrl
 	int fd, ret;
 	uint32_t msg;
 
+	printf("connect_fork\n");
 	fd = fd_getd(socket);
 	ret = real.connect(fd, addr, addrlen);
+	printf("connect_fork - real connect %d\n", ret);
 	if (ret)
 		return ret;
 
 	ret = real.recv(fd, &msg, sizeof msg, MSG_PEEK);
+	printf("connect_fork - real recv %d msg %d\n", ret, msg);
 	if ((ret != sizeof msg) || msg) {
 		fd_store(socket, fd, fd_normal);
 		return 0;
 	}
 
 	ret = transpose_socket(socket, fd_rsocket);
+	printf("connect_fork - transpose socket %d\n", ret);
 	if (ret < 0)
 		return ret;
 
@@ -830,11 +835,15 @@ pid_t fork(void)
 	uint32_t msg;
 
 	init_preload();
+	printf("fork\n");
 	pid = real.fork();
+	printf("fork - pid %d fork_support %d last_accept %d \n",
+		pid, fork_support, last_accept);
 	if (pid || !fork_support || (last_accept < 0) ||
 	    (fd_get(last_accept, &sfd) != fd_fork))
 		goto out;
 
+	printf("fork - switching to rsocket\n");
 	len = sizeof sin6;
 	ret = real.getsockname(sfd, (struct sockaddr *) &sin6, &len);
 	if (ret)
@@ -843,10 +852,12 @@ pid_t fork(void)
 	memset(&sin6.sin6_addr, 0, sizeof sin6.sin6_addr);
 
 	sem = sem_open("/rsocket_fork", O_CREAT, 0644, 1);
+	printf("fork - sem_open\n");
 	if (sem == SEM_FAILED)
 		goto out;
 
 	lfd = rsocket(sin6.sin6_family, SOCK_STREAM, 0);
+	printf("fork - rsocket %d\n", lfd);
 	if (lfd < 0)
 		goto sclose;
 
@@ -855,14 +866,17 @@ pid_t fork(void)
 
 	sem_wait(sem);
 	ret = rbind(lfd, (struct sockaddr *) &sin6, sizeof sin6);
+	printf("fork - rbind %d\n", ret);
 	if (ret)
 		goto lclose;
 
 	ret = rlisten(lfd, 1);
+	printf("fork - rlisten %d\n", ret);
 	if (ret)
 		goto lclose;
 
 	dfd = raccept(lfd, NULL, NULL);
+	printf("fork - raccept %d\n", dfd);
 	if (dfd < 0)
 		goto lclose;
 
@@ -872,6 +886,7 @@ pid_t fork(void)
 
 	msg = 0;
 	ret = real.write(sfd, &msg, sizeof msg);
+	printf("fork - real write %d\n", ret);
 	if (ret != sizeof msg) {
 		rclose(dfd);
 		goto lclose;
